src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +121,235 @@ namespace wasm {
 
 namespace {
 
+// Contains information about a RefCast we want to move to some target.
+struct RefCastInfo {
+  Type type;
+  RefCast::Safety safety;
+
+  // Indicate whether the cast is already at the target. If it is, then
+  // the information is just for determining if a potential cast is less
+  // refined than the existing cast, and will be removed later.
+  bool atTarget;
+
+  RefCastInfo() {}
+
+  RefCastInfo(Type type, RefCast::Safety safety, bool atTarget)
+    : type(type), safety(safety), atTarget(atTarget) {}
+};
+
+// Find a cast to move earlier to another local.get. More refined subtypes are
+// chosen over less refined ones.
+struct EarlyCastFinder
+  : public LinearExecutionWalker<EarlyCastFinder,
+                                 UnifiedExpressionVisitor<EarlyCastFinder>> {
+  PassOptions options;
+  size_t numLocals;
+
+  // Tracks the current earliest local.get that we can move a cast to without
+  // side-effects.
+  std::vector<LocalGet*> earliestRefCastReachable;
+  std::vector<LocalGet*> earliestRefAsReachable;
```

Comment:
@kripken yes, `atTarget` would not be necessary, but wouldn't it save some extra `getFallthrough` calls?

Reply:
Yes, it's a tradeoff. I'm actually not sure how the tradeoff works in terms of runtime performance, since I can see benefits both ways (more `getFallthrough` calls, as you said, vs. smaller data structures). Given that, I think smaller and clearer code seems best.

Reply:
Changes made.